File: skills/install-slop-guard/assets/typescript/smells/rules/no-primitive-obsession.ts
import { defineRule } from "@oxlint/plugins";
import type { ESTree } from "@oxlint/plugins";

import { numberOption, thresholdSchema } from "../shared/options.ts";

const DEFAULT_MAX_SAME_TYPE = 2;

const PRIMITIVE_KEYWORDS: Readonly<Record<string, string>> = {
	TSStringKeyword: "string",
	TSNumberKeyword: "number",
	TSBooleanKeyword: "boolean",
	TSBigIntKeyword: "bigint",
};

type ParameterOwner =
	| ESTree.ArrowFunctionExpression
	| ESTree.Function
	| ESTree.TSMethodSignature;

function parameterName(parameter: ESTree.ParamPattern): string | null {
	if (parameter.type === "Identifier") return parameter.name;
	if (parameter.type === "TSParameterProperty") return parameterName(parameter.parameter);
	return null;
}

/** Ban runs of same-typed primitive parameters that a domain type should carry. */
export const noPrimitiveObsessionRule = defineRule({
	meta: {
		type: "problem",
		docs: {
			description:
				"Disallow more than `maxSameType` parameters of the same primitive type; give the domain concept a type of its own.",
		},
		messages: {
			primitiveObsession:
				"This function takes {{count}} `{{primitive}}` parameters ({{names}}), so any two of them can be swapped at a call site and nothing complains. Give the values a type of their own.",
		},
		...thresholdSchema({ maxSameType: DEFAULT_MAX_SAME_TYPE }),
	},
	createOnce(context) {
		const check = (node: ParameterOwner) => {
			const limit = numberOption(context.options, "maxSameType", DEFAULT_MAX_SAME_TYPE);
			const counts = new Map<string, string[]>();
			for (const parameter of node.params) {
				const name = parameterName(parameter);
				const annotation =
					parameter.type === "Identifier" ? parameter.typeAnnotation?.typeAnnotation : undefined;
				if (name === null || annotation === undefined || annotation === null) continue;
				const primitive = PRIMITIVE_KEYWORDS[annotation.type];
				if (primitive === undefined) continue;
				counts.set(primitive, [...(counts.get(primitive) ?? []), name]);
			}
			for (const [primitive, names] of counts) {
				if (names.length <= limit) continue;
				context.report({
					node,
					messageId: "primitiveObsession",
					data: {
						count: String(names.length),
						primitive,
						names: names.map((name) => `\`${name}\``).join(", "),
					},
				});
			}
		};

		return {
			ArrowFunctionExpression: check,
			FunctionDeclaration: check,
			FunctionExpression: check,
			TSDeclareFunction: check,
			TSMethodSignature: check,
		};
	},
});

File: skills/install-slop-guard/assets/python/rules/smells/no_primitive_obsession.py
"""Reject runs of same-typed primitive parameters that a domain type should carry."""

from __future__ import annotations

import ast

from ...rule import FunctionNode, Rule

PRIMITIVES = frozenset({"str", "int", "float", "bool", "bytes"})
DEFAULT_MAX_SAME_TYPE = 2
RECEIVERS = frozenset({"self", "cls", "mcs"})


class NoPrimitiveObsession(Rule):
    name = "no-primitive-obsession"
    group = "smells"
    reference = "https://refactoring.guru/smells/primitive-obsession"
    description = (
        "Disallow more than `max_same_type` parameters of the same primitive type, 2 by default; "
        "give the domain concept a type."
    )

    def visit_FunctionDef(self, node: FunctionNode) -> None:
        limit = self.threshold("max_same_type", DEFAULT_MAX_SAME_TYPE)
        counts: dict[str, list[str]] = {}
        for parameter in self.parameters(node):
            if parameter.arg in RECEIVERS:
                continue
            annotation = parameter.annotation
            if isinstance(annotation, ast.Name) and annotation.id in PRIMITIVES:
                counts.setdefault(annotation.id, []).append(parameter.arg)
        for primitive, names in counts.items():
            if len(names) > limit:
                listed = ", ".join(f"`{name}`" for name in names)
                self.report(
                    node,
                    f"`{node.name}` takes {len(names)} `{primitive}` parameters ({listed}), so "
                    f"any two of them can be swapped at a call site and nothing complains. Give the "
                    f"values a type of their own.",
                )
        self.generic_visit(node)

    visit_AsyncFunctionDef = visit_FunctionDef
